### DISCUSSION TITLE
How to render both side of a mesh?

### DISCUSSION THREAD

**Discussion by Galaxeaaa** [mitsuba-renderer/mitsuba3#1107] [Q&A] [answered]
Hi, I'm new to Mitsuba and I know this is a stupid question, but I really haven't found a convenient way to render the mesh on either side of it. The back is always black. In this render, for example, the floor's normal points downward.
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/61783063/969279bd-050e-4ce9-86b0-b23f0bfd0425)

Thank you!

**Comment by ziyi-zhang** [comment, marked as answer]
Hi @Galaxeaaa,

That can be achieved using a double sided BSDF. The doc is [here](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#two-sided-brdf-adapter-twosided).
Best

**Reply by Galaxeaaa** [reply]
Hi @ziyi-zhang,

Thank you so much! I missed this type of BSDF and it works.

I actually have another related question now: Since I want all the BSDFs to be double sided in the XML file, manually adding this wrapper is tedious. Is writing a custom BSDF plugin the only way to do that?

And another problem I encountered recently: How can I scale a texture in XML file?

Thank you!

**Reply by merlinND** [reply]
Hello @Galaxeaaa,

To answer your first question, a simpler way might be to programatically generate your scene from Python. This way, you can easily wrap any BSDF withing a `twosided` BSDF. The dict-based Python representation is equivalent to the XML scene format. The documentation contains examples of both XML and Python-style scene formats.

If you define your scene in Python, you will also be able to load the texture and process it arbitrarily before passing the processed Bitmap object directly into the `dict` that describes your texture ([`bitmap` field](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_textures.html#bitmap-texture-bitmap))

**Reply by Galaxeaaa** [reply]
Hi @merlinND,

Since I'm using a dataset providing xml files, I'll need to convert xml to Python `dict` first in this case. I wonder if there's a built-in function to do this? I only found `mitsuba.xml_to_props` [here](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.xml_to_props).

Thanks!

**Reply by ziyi-zhang** [reply]
No we do not provide such a conversion. Two alternative solutions:
- you can (ask chatgpt to) write a small python script to generate the double-sided version of xml
- or load the scene, and traverse the parameters using "mi.traverse(scene)" to replace the bsdfs with their double sided version.
Best

**Reply by Galaxeaaa** [reply]
I've written a script to generate new xml already, but I feel the second one is even better. Thank you so much for your suggestion!

**Reply by ziyi-zhang** [reply]
Hi @Galaxeaaa 
As reminded by merlinND, this actually requires changing the type of BSDF, which is not exposed in the traversal -- the traversal only exposes parameters of the current BSDF, not the BSDF itself.
Please ignore the second suggestion. Sorry for any inconvenience.

**Reply by Galaxeaaa** [reply]
I see. Thanks!
